File: adapters/dart/lib/src/humanize.dart
// Deterministic identifier-rule humanization (design §8.1, provenance
// 'derived'). The full Korean naming engine lands in ticket 11; this module
// only does the English fallback: split camelCase/snake_case words and join
// them as an English sentence fragment, e.g.
//   submitOrder          -> "Submit order"
//   _onItemAdded         -> "Item added"
//   onCheckoutPressed    -> "Checkout pressed"
//   firebaseMessagingBackgroundHandler
//                        -> "Firebase messaging background handler"
//   URLLoader            -> "Url loader"

bool _isUpperAt(String s, int i) {
  if (i < 0 || i >= s.length) return false;
  final ch = s.substring(i, i + 1);
  return ch.toUpperCase() == ch && ch.toLowerCase() != ch;
}

bool _isDigitAt(String s, int i) {
  if (i < 0 || i >= s.length) return false;
  final c = s.codeUnitAt(i);
  return c >= 0x30 && c <= 0x39;
}

/// Splits [s] on underscores, dollar signs and camelCase boundaries.
/// Consecutive uppercase letters stay together until an acronym run ends:
/// `URLLoader` -> `[URL, Loader]`.
List<String> splitIdentifierWords(String s) {
  final words = <String>[];
  final current = StringBuffer();
  void flush() {
    if (current.isNotEmpty) {
      words.add(current.toString());
      current.clear();
    }
  }

  for (var i = 0; i < s.length; i++) {
    final ch = s[i];
    if (ch == '_' || ch == r'$') {
      flush();
      continue;
    }
    if (_isUpperAt(s, i)) {
      // camel boundary: "fooBar"/"foo9Bar" -> foo | Bar
      final afterLowerOrDigit =
          current.isNotEmpty && (_isLowerAt(s, i - 1) || _isDigitAt(s, i - 1));
      // acronym-run end: "URLLoader" at 'L' -> URL | Loader
      final acronymEnd =
          current.length >= 2 && _isUpperAt(s, i - 1) && _isLowerAt(s, i + 1);
      if (afterLowerOrDigit || acronymEnd) {
        flush();
      }
    }
    current.write(ch);
  }
  flush();
  return words;
}

bool _isLowerAt(String s, int i) {
  if (i < 0 || i >= s.length) return false;
  final ch = s.substring(i, i + 1);
  return ch.toLowerCase() == ch && ch.toUpperCase() != ch;
}

/// Humanizes a Dart symbol name into a short English phrase. Always returns a
/// non-empty string (falls back to the trimmed raw name).
String humanizeIdentifier(String rawName) {
  var name = rawName.trim();
  while (name.startsWith('_')) {
    name = name.substring(1);
  }
  // Drop a leading event-handler "on" prefix when followed by an uppercase
  // letter: onItemAdded -> ItemAdded.
  if (name.length > 2 &&
      name.startsWith('on') &&
      _isAsciiUpper(name.codeUnitAt(2))) {
    name = name.substring(2);
  }
  final words = splitIdentifierWords(name)
      .map((w) => w.toLowerCase())
      .where((w) => w.isNotEmpty)
      .toList();
  if (words.isEmpty) {
    return 'unnamed';
  }
  final first = words.first;
  final buf = StringBuffer()
    ..write(first.substring(0, 1).toUpperCase())
    ..write(first.substring(1));
  for (var i = 1; i < words.length; i++) {
    buf.write(' ');
    buf.write(words[i]);
  }
  return buf.toString();
}

bool _isAsciiUpper(int codeUnit) => codeUnit >= 0x41 && codeUnit <= 0x5A;

File: adapters/typescript/lib/humanize.js
'use strict';

/**
 * Splits an identifier on camelCase, snake_case, underscores, and dollar signs.
 * Handles acronym runs properly: e.g. "URLLoader" -> ["URL", "Loader"].
 * @param {string} raw
 * @returns {string[]}
 */
function splitIdentifierWords(raw) {
  let s = String(raw || '').trim().replace(/^_+/, '');
  // Drop leading "on" prefix when followed by an uppercase letter: onItemAdded -> ItemAdded
  if (s.length > 2 && s.startsWith('on') && s[2] >= 'A' && s[2] <= 'Z') {
    s = s.substring(2);
  }

  const words = [];
  let current = '';

  for (let i = 0; i < s.length; i++) {
    const ch = s[i];
    if (ch === '_' || ch === '$') {
      if (current.length > 0) {
        words.push(current);
        current = '';
      }
      continue;
    }

    const isUpper = ch >= 'A' && ch <= 'Z';
    const isDigit = ch >= '0' && ch <= '9';
    const prevLowerOrDigit = i > 0 && ((s[i - 1] >= 'a' && s[i - 1] <= 'z') || (s[i - 1] >= '0' && s[i - 1] <= '9'));
    const isAcronymEnd = current.length >= 2 && s[i - 1] >= 'A' && s[i - 1] <= 'Z' && i + 1 < s.length && s[i + 1] >= 'a' && s[i + 1] <= 'z';

    if (isUpper && (prevLowerOrDigit || isAcronymEnd)) {
      if (current.length > 0) {
        words.push(current);
        current = '';
      }
    }
    current += ch;
  }

  if (current.length > 0) {
    words.push(current);
  }

  return words;
}

/**
 * Humanizes an identifier into a clean English sentence fragment.
 * e.g. "submitOrder" -> "Submit order"
 * @param {string} rawName
 * @returns {string}
 */
function humanizeIdentifier(rawName) {
  const words = splitIdentifierWords(rawName)
    .map(w => w.toLowerCase())
    .filter(w => w.length > 0);

  if (words.length === 0) {
    return 'Unnamed';
  }

  const first = words[0];
  const capitalizedFirst = first.charAt(0).toUpperCase() + first.slice(1);
  words[0] = capitalizedFirst;

  return words.join(' ');
}

module.exports = {
  splitIdentifierWords,
  humanizeIdentifier,
};
